bitcoin/BRSyncManager.c

```diff
@@ -1820,3 +1883,127 @@ _BRPeerSyncManagerTxPublished (void *info,
+    array_new (newAddresses, addressCount);
+
+    for (size_t index = 0; index < addressCount; index++) {
+        if (!BRSetContains (addresses, &addressArray[index])) {
+            // one copy remains owned by the address set
+            BRAddress *address = malloc (sizeof(BRAddress));
+            *address = addressArray[index];
+            BRSetAdd (addresses, address);
+
+            // one copy owned by the returned array
+            address = malloc (sizeof(BRAddress));
+            *address = addressArray[index];
+            array_add (newAddresses, address);
+        }
+    }
+
+    free (addressArray);
+    return newAddresses;
+}
+
+static uint32_t
+_getLastConfirmedSendTxHeight(BRWallet *wallet,
+                              uint64_t lastBlockHeight,
+                              uint64_t confirmationsUntilFinal) {
+    uint32_t scanHeight = 0;
+
+    size_t transactionCount = BRWalletTransactions (wallet, NULL, 0);
+
+    if (lastBlockHeight >= confirmationsUntilFinal && 0 != transactionCount) {
+        BRArrayOf(BRTransaction *) transactions = (BRArrayOf(BRTransaction *) ) calloc (transactionCount, sizeof (BRTransaction *));
+        transactionCount = BRWalletTransactions (wallet, transactions, transactionCount);
+
+        for (size_t index = 0; index < transactionCount; index++) {
+            // ensure:
+            // - tx is valid (i.e. no previous transaction spend any of utxos, and no inputs are invalid)
+            // - AND the transaction was a SEND
+            // - AND the transaction has been confirmed
+            if (BRWalletTransactionIsValid (wallet, transactions[index]) &&
+                0 != BRWalletAmountSentByTx (wallet, transactions[index]) &&
+                TX_UNCONFIRMED != transactions[index]->blockHeight &&
+                transactions[index]->blockHeight < (lastBlockHeight - confirmationsUntilFinal)) {
+                scanHeight = (transactions[index]->blockHeight > scanHeight ?
+                                transactions[index]->blockHeight :
+                                scanHeight);
+            }
+        }
+
+        free (transactions);
+    }
+
+    return scanHeight;
+}
+
+static uint32_t
+_calculateSyncDepthHeight(BRSyncDepth depth,
+                          BRWallet *wallet,
+                          const BRChainParams *chainParams,
+                          uint64_t networkBlockHeight,
+                          uint64_t confirmationsUntilFinal) {
+    uint32_t scanHeight = 0;
+
+    switch (depth) {
+        case SYNC_DEPTH_FROM_LAST_CONFIRMED_SEND: {
+            scanHeight = _getLastConfirmedSendTxHeight (wallet,
+                                                        networkBlockHeight,
+                                                        confirmationsUntilFinal);
+            break;
+        }
+        case SYNC_DEPTH_FROM_LAST_TRUSTED_BLOCK: {
+            const BRCheckPoint *checkpoint = BRChainParamsGetCheckpointBeforeBlockNumber (chainParams,
```

Comment:
Does it make sense to use checkpoints when in API mode? I don't think there is any harm personally.

Reply:
I'd do it as it parallels P2P mode as one less thing: to document, to explain to 'customer service' about recovery, to account for in test event/log sequences.